src/qutip_qip/circuit.py

```diff
@@ -202,7 +202,7 @@ def __init__(self, N, input_states=None, output_states=None,
         self.N = N
         self.reverse_states = reverse_states
         self.gates = []
-        self.dims = dims
+        self.dims = dims if dims is not None else [2] * N
```

Comment:
Question for my own understanding: What are the cases where it would be useful to have dims not be `[2] * N`?

Reply:
In principle, there is nothing preventing one from defining a qutrit/qudit/multi_level circuit. Although there is no predefined gate for that. It is not very popular at the moment but it is also not very hard to support it.

Reply:
Woot. Maybe we should write some tests or examples with qutrit+ gates at some point so that we can be sure it all works and so that maybe users can find it and try it out.

Reply:
There is one :)

https://github.com/qutip/qutip-qip/blob/8b49a89405183eb3e264abfc50aab3bf47fe2ef8/tests/test_circuit.py#L513-L532

Reply:
It is just that the *default* `dims` was not used at all before because we don't have predefined qutrits gate anyway. But now I'm using it in expanding the circuit.

But you are right that it is not tested if part of the circuit is qubits using the predefined gate, while the rest is using custom qutrits gate.